### PR TITLE
[blip_2] Support attn_implementation dispatch

### PR DESCRIPTION
**Pull request by YangKai0616**
# What does this PR do?

As per the title.

### CODE REVIEW

**Comment by YangKai0616** [comment]
@vasqu

File: src/transformers/models/blip_2/modeling_blip_2.py [hunk 1]
```diff
@@ -881,9 +852,7 @@ def forward(
 class Blip2QFormerModel(Blip2PreTrainedModel):
     config: Blip2QFormerConfig
 
-    _supports_attention_backend = False  # adds position on attn weights before last matmul
```
Comment (zucchini-nlp) [inline src/transformers/models/blip_2/modeling_blip_2.py:884]: have to set it explicitly to True here and in other models

File: src/transformers/models/blip_2/modeling_blip_2.py [hunk 2]
```diff
@@ -881,9 +852,7 @@ def forward(
...
     _supports_flash_attn = False
```
Comment (zucchini-nlp) [inline src/transformers/models/blip_2/modeling_blip_2.py]: i think we can do FA and flex now, no?

File: src/transformers/models/blip_2/modeling_blip_2.py [hunk 3]
```diff
@@ -591,49 +573,38 @@ def forward(
...
-
-        # Take the dot product between "query" and "key" to get the raw attention scores.
-        attention_scores = torch.matmul(query_layer, key_layer.transpose(-1, -2))
-
-        attention_scores = attention_scores / math.sqrt(self.attention_head_size)
-
-        if attention_mask is not None:
-            # Apply the attention mask is (precomputed for all layers in BertModel forward() function)
-            attention_scores = attention_scores + attention_mask
+            current_states = hidden_states
```
Comment (zucchini-nlp) [inline src/transformers/models/blip_2/modeling_blip_2.py:583]: ohh nice, i forgot we got rid of those `position_embeddings` which weren't used by official ckpt

File: src/transformers/models/blip_2/modeling_blip_2.py (hunk 2 above)
Comment (YangKai0616) [reply src/transformers/models/blip_2/modeling_blip_2.py]: FA still looks blocked by the `QFormer` fp32 path, flex does not seem blocked in the same way as FA, but flex fails with attention dropout in training, and `QFormer` attention recording expects attention weights while flex returns `LSE` instead. Perhaps we should keep this PR to SDPA content only?👀

File: src/transformers/models/blip_2/modeling_blip_2.py (hunk 1 above)
Comment (YangKai0616) [reply src/transformers/models/blip_2/modeling_blip_2.py:884]: Done.

**Comment by YangKai0616** [comment]
For the test case `test_modeling_instructblip.py::InstructBlipModelIntegrationTest::test_inference_flant5_xl`, the output from the current default SDPA branch differs from the default eager mode in upstream/main. 
However, it matches when using eager mode on the current branch or float32 dtype. `Expectations` should be updated, but since I don't know the device used in your CI, this PR will not be updated for now.

File: src/transformers/models/blip_2/modeling_blip_2.py [hunk 4]
```diff
@@ -558,26 +560,6 @@ def __init__(self, config, is_cross_attention=False):
...
-    def save_attention_map(self, attention_map):
-        self.attention_map = attention_map
-
-    def get_attention_map(self):
-        return self.attention_map
-
-    def transpose_for_scores(self, x):
-        new_x_shape = x.size()[:-1] + (self.num_attention_heads, self.attention_head_size)
-        x = x.view(*new_x_shape)
-        return x.permute(0, 2, 1, 3)
```
Comment (vasqu) [inline src/transformers/models/blip_2/modeling_blip_2.py:579]: Oh god that was ugly 😬 glad to get rid of this

File: src/transformers/models/blip_2/modeling_blip_2.py [hunk 5]
```diff
@@ -881,9 +852,9 @@ def forward(
 class Blip2QFormerModel(Blip2PreTrainedModel):
     config: Blip2QFormerConfig
 
-    _supports_attention_backend = False  # adds position on attn weights before last matmul
-    _supports_flash_attn = False
-    _supports_sdpa = False
+    _supports_attention_backend = True
+    _supports_sdpa = True
+    _supports_flash_attn = False  # Q-Former is kept in fp32, which blocks reliable Flash Attention dispatch.
     _supports_flex_attn = False
```
Comment (vasqu) [inline src/transformers/models/blip_2/modeling_blip_2.py]: Flex attention should work no?

File: src/transformers/models/blip_2/modeling_blip_2.py [hunk 6]
```diff
@@ -881,9 +852,9 @@ def forward(
 class Blip2QFormerModel(Blip2PreTrainedModel):
     config: Blip2QFormerConfig
 
-    _supports_attention_backend = False  # adds position on attn weights before last matmul
-    _supports_flash_attn = False
-    _supports_sdpa = False
+    _supports_attention_backend = True
+    _supports_sdpa = True
+    _supports_flash_attn = False  # Q-Former is kept in fp32, which blocks reliable Flash Attention dispatch.
```
Comment (vasqu) [inline src/transformers/models/blip_2/modeling_blip_2.py:857]: Oh yea I remember that one...

File: src/transformers/models/instructblip/modeling_instructblip.py [hunk 7]
```diff
@@ -791,9 +764,9 @@ class InstructBlipQFormerModel(InstructBlipPreTrainedModel):
     instruction as input.
     """
 
-    _supports_attention_backend = False  # adds position on attn weights before last matmul
+    _supports_attention_backend = True
+    _supports_sdpa = True
     _supports_flash_attn = False
```
Comment (vasqu) [inline src/transformers/models/instructblip/modeling_instructblip.py]: Lets add comments when why not

Comment (vasqu) [reply src/transformers/models/instructblip/modeling_instructblip.py]: And same re flex

File: src/transformers/models/instructblipvideo/modeling_instructblipvideo.py [hunk 8]
```diff
@@ -740,9 +713,9 @@ class InstructBlipVideoQFormerModel(InstructBlipVideoPreTrainedModel):
     instruction as input.
     """
 
-    _supports_attention_backend = False  # adds position on attn weights before last matmul
+    _supports_attention_backend = True
+    _supports_sdpa = True
     _supports_flash_attn = False
-    _supports_sdpa = False
     _supports_flex_attn = False
```
Comment (vasqu) [inline src/transformers/models/instructblipvideo/modeling_instructblipvideo.py]: Same here

File: tests/models/blip_2/test_modeling_blip_2.py [hunk 9]
```diff
@@ -506,6 +500,11 @@ def test_generate_without_input_ids(self):
     def test_model_base_model_prefix(self):
         pass
 
+    def test_sdpa_can_dispatch_on_flash(self):
```
Comment (vasqu) [inline tests/models/blip_2/test_modeling_blip_2.py:516]: yes but lets use the `@unittest.skip` decorator please

File: src/transformers/models/blip_2/modeling_blip_2.py [hunk 10]
```diff
@@ -549,6 +548,9 @@ def __init__(self, config, is_cross_attention=False):
         self.num_attention_heads = config.num_attention_heads
         self.attention_head_size = int(config.hidden_size / config.num_attention_heads)
         self.all_head_size = self.num_attention_heads * self.attention_head_size
+        self.scaling = self.attention_head_size**-0.5
+        self.is_causal = False
+        self.attention_dropout = config.attention_probs_dropout_prob
```
Comment (vasqu) [inline src/transformers/models/blip_2/modeling_blip_2.py:553]:
```suggestion
        self.dropout = config.attention_probs_dropout_prob
```
nit

**Review by vasqu — COMMENTED** [review]
Just a few smaller comments but overall looks good already

**Comment by vasqu** [comment]
> For the test case test_modeling_instructblip.py::InstructBlipModelIntegrationTest::test_inference_flant5_xl, the output from the current default SDPA branch differs from the default eager mode in upstream/main.
However, it matches when using eager mode on the current branch or float32 dtype. Expectations should be updated, but since I don't know the device used in your CI, this PR will not be updated for now.

Can we force eager attention at load time instead for now?

File: src/transformers/models/blip_2/modeling_blip_2.py (hunk 5 above)
Comment (YangKai0616) [reply src/transformers/models/blip_2/modeling_blip_2.py]: Yes, updated here.

File: src/transformers/models/instructblip/modeling_instructblip.py (hunk 7 above)
Comment (YangKai0616) [reply src/transformers/models/instructblip/modeling_instructblip.py]: Previously unverified. Verified now, updated the flag and related tests.

File: src/transformers/models/instructblipvideo/modeling_instructblipvideo.py (hunk 8 above)
Comment (YangKai0616) [reply src/transformers/models/instructblipvideo/modeling_instructblipvideo.py]: Done.

File: tests/models/blip_2/test_modeling_blip_2.py (hunk 9 above)
Comment (YangKai0616) [reply tests/models/blip_2/test_modeling_blip_2.py:516]: Done.

File: src/transformers/models/blip_2/modeling_blip_2.py (hunk 10 above)
Comment (YangKai0616) [reply src/transformers/models/blip_2/modeling_blip_2.py:553]: Done.

**Comment by YangKai0616** [comment]
> Can we force eager attention at load time instead for now?

Done.

File: src/transformers/models/blip_2/modeling_blip_2.py [hunk 11]
```diff
@@ -296,7 +295,7 @@ def __init__(self, config):
             )
         self.scale = self.head_dim**-0.5
         self.is_causal = False
-        self.attention_dropout = config.attention_dropout
+        self.dropout = config.attention_dropout
```
Comment (vasqu) [inline src/transformers/models/blip_2/modeling_blip_2.py]: Ah I was missing this context previously: You named it attention_dropout because of this attribute naming here. This makes more sense then. Let's revert this back to attention dropout then (otherwise we are slightly breaking)

File: tests/models/blip_2/test_modeling_blip_2.py [hunk 12]
```diff
@@ -1424,6 +1438,16 @@ def test_retain_grad_hidden_states_attentions(self):
...
         pass
 
+    @unittest.skip(
+        reason="QFormer is forced to fp32 via _keep_in_fp32_modules, incompatible with SDPA flash-only kernel"
+    )
+    def test_sdpa_can_dispatch_on_flash(self):
+        pass
+
+    @unittest.skip(reason="_prepare_config_headdim breaks the qformer/vision cross-attention hidden-size invariant")
+    def test_flex_attention_with_grads(self):
```
Comment (vasqu) [inline tests/models/blip_2/test_modeling_blip_2.py]: Could we fix that with a small override?

File: tests/models/instructblip/test_modeling_instructblip.py [hunk 13]
```diff
@@ -528,6 +525,130 @@ def test_generate_without_input_ids(self):
...
+
+    @unittest.skip(reason="_prepare_config_headdim breaks the qformer/vision cross-attention hidden-size invariant")
+    def test_flash_attn_2_inference_equivalence(self):
+        pass
+
+    @unittest.skip(reason="_prepare_config_headdim breaks the qformer/vision cross-attention hidden-size invariant")
+    def test_flash_attn_2_inference_equivalence_right_padding(self):
+        pass
+
+    @unittest.skip(reason="_prepare_config_headdim breaks the qformer/vision cross-attention hidden-size invariant")
```
Comment (vasqu) [inline tests/models/instructblip/test_modeling_instructblip.py]: same here, I think we should rather override `_prepare_config_headdim` to account for these models

**Review by vasqu — APPROVED** [review]
Just some smaller things but overall ready. Biggest thing to fix is imo to properly override `_prepare_config_headdim` instead of skipping as this affects the whole flex and FA test suite

File: src/transformers/models/blip_2/modeling_blip_2.py (hunk 11 above)
Comment (YangKai0616) [reply src/transformers/models/blip_2/modeling_blip_2.py]: Done.

File: tests/models/instructblip/test_modeling_instructblip.py (hunk 13 above)
Comment (YangKai0616) [reply tests/models/instructblip/test_modeling_instructblip.py]: Done.

**Comment by YangKai0616** [comment]
Friendly ping @vasqu

File: src/transformers/models/instructblip/modeling_instructblip.py [hunk 14]
```diff
@@ -858,9 +831,10 @@ def forward(
         )
 
         if encoder_attention_mask is not None:
+            query_embedding_output = embedding_output[:, :query_length, :] if query_length > 0 else embedding_output
```
Comment (vasqu) [inline src/transformers/models/instructblip/modeling_instructblip.py:836]: Oh where did you find this issue arising? Ig this is the issue potentially on cross attn?

File: tests/models/instructblip/test_modeling_instructblip.py [hunk 15]
```diff
@@ -528,6 +536,135 @@ def test_generate_without_input_ids(self):
...
+            super().flash_attn_inference_equivalence(
+                attn_implementation=attn_implementation, padding_side=padding_side, atol=atol, rtol=rtol
+            )
+
+    @require_flash_attn
+    @require_torch_accelerator
+    @require_bitsandbytes
+    @pytest.mark.flash_attn_test
+    @slow
+    def test_flash_attn_2_fp32_ln(self):
```
Comment (vasqu) [inline tests/models/instructblip/test_modeling_instructblip.py:577]: Can we add small comments for the overrides as to why?

Comment (vasqu) [reply tests/models/instructblip/test_modeling_instructblip.py:577]: Same for below

**Review by vasqu — APPROVED** [review]
LGTM, just see my last comments but nothing major

File: src/transformers/models/instructblip/modeling_instructblip.py (hunk 14 above)
Comment (YangKai0616) [reply src/transformers/models/instructblip/modeling_instructblip.py:836]: Yes, it's on the cross-attention path. The Q-Former's cross-attention queries come only from the query tokens, but `embedding_output` also contains the text tokens. `create_bidirectional_mask` infers the query dimension from `inputs_embeds`, so passing the full `embedding_output` produced a mask whose query dim didn't match the actual cross-attention query length. I found it in the error of the test case.

Comment (YangKai0616) [reply src/transformers/models/instructblip/modeling_instructblip.py:836]: Comments added.

File: tests/models/instructblip/test_modeling_instructblip.py (hunk 15 above)
Comment (YangKai0616) [reply tests/models/instructblip/test_modeling_instructblip.py:577]: Done.

**Comment by vasqu** [comment]
run-slow: blip_2, instructblip, instructblipvideo

**Comment by vasqu** [comment]
Just checking with slow CI but merging afterwards :hugs:
